### PR TITLE
'stack ghci': build project before launching GHCi

### LINKED ISSUE #1065: Stack ghci before stack build fails
C:\Neil\hlint>stack ghci
    Using main module: Package `hlint' component exe:hlint with main-is file: C:\Neil\hlint\src\Main.hs
    Configuring GHCi with the following packages: hlint
    GHCi, version 7.10.2: http://www.haskell.org/ghc/  :? for help

    <command line>:
        Could not find module `Paths_hlint'
        Use -v to see a list of the files searched for.
    Failed, modules loaded: none.
    Prelude> :q
    Leaving GHCi.

That's after deleting my .ghci file and removing my .stack-work.

My guess is `stack ghci` should generate the paths file if it needs it. I thought it used to do a whole build before launching ghci? Using 06a4b270de2d2dcbd3ee8abe98078151c4a8f075.

### LINKED ISSUE #1180: 'stack ghci' doesn't notice that a module got deleted
If you remove a module and then run `stack ghci` before `stack build`, it fails with a `Could not find module` error.

Hypothesis:

The build doesn't run before `stack ghci` reads the modules, so it doesn't get a chance to re-write the .dump-hi files with the updated information (that the module does not exist), so `stack ghci` is still reading the out-of-date information.

Notes from @chrisdone:

> as a heads up the stack ghci code uses the build info, but runs the equiv. of `stack build —only-dependencies` to ensure that dependencies are available
> at a guess, it seems that the build cache doesn’t get invalidated by simply reading it in the manner that stack ghci does, but i didn’t get niklas to test with `stack build` to see if running that would trigger a cache invalidation
> the building with `—only-dependencies` was enabled to avoid ghci not being able to start when there’s a build error. i don’t like that fix, i’d prefer if the build just ran and if it fails then it should catch the exception and print the compile error but still launch ghci anyway in an optimistic way (edited)

It's likely that fixing #1065 will also fix this, but tracking separately since the symptoms are different.

(PAT-125)

### PR DESCRIPTION
**Pull request by borsboom**
(closes #1065, fixes #1180)

Optimistically, launch GHCi anyway even if the build fails.

@chrisdone, can you review?

### CODE REVIEW

File: src/Stack/Ide.hs [hunk 1]
```diff
@@ -43,7 +43,7 @@ ide
     -> [String] -- ^ GHC options.
     -> m ()
 ide targets useropts = do
-    (_realTargets,_,pkgs) <- ghciSetup Nothing targets
+    (_realTargets,_,pkgs) <- ghciSetup (Just BSAll) Nothing targets
```
Comment (borsboom) [inline src/Stack/Ide.hs]: @chrisdone @mgsloan: Is building the project (and continuing even if it fails) the correct behaviour for `stack ide`?

File: src/main/Main.hs [hunk 2]
```diff
@@ -884,7 +884,7 @@ packagesCmd () go@GlobalOpts{..} =
 targetsCmd :: Text -> GlobalOpts -> IO ()
 targetsCmd target go@GlobalOpts{..} =
     withBuildConfig go $
-    do (_realTargets,_,pkgs) <- ghciSetup Nothing [target]
+    do (_realTargets,_,pkgs) <- ghciSetup (Just BSAll) Nothing [target]
```
Comment (borsboom) [inline src/main/Main.hs]: @chrisdone: Is building the project (and continuing even if it fails) the correct behaviour for `stack targets`?

**Comment by borsboom** [comment]
I've rebased this to resolve merge conflicts.  @chrisdone / @mgsloan: still not sure what the right behaviour for `stack ide` and `stack targets` should be.

File: src/main/Main.hs (hunk 2 above)
Comment (mgsloan) [inline src/main/Main.hs]: Not building at all is fine for the targets command.

File: src/Stack/Ide.hs (hunk 1 above)
Comment (mgsloan) [inline src/Stack/Ide.hs]: Actually, I think the right behavior for stack ide is to only build dependencies.

**Comment by mgsloan** [comment]
LGTM!  should the `--dependencies-only` alias be part of the BuildSubset parser?

File: src/Stack/Ghci.hs [hunk 3]
```diff
@@ -215,12 +218,22 @@ ghciSetup mainIs stringTargets = do
                                  return (Just (name, (cabalfp, simpleTargets)))
                              Nothing -> return Nothing
                     else return Nothing
+    -- Try to build, but optimistically launch GHCi anyway if it fails (#1065)
+    case (mbuildFirst, M.null realTargets) of
```
Comment (mgsloan) [inline src/Stack/Ghci.hs]: I think the check for (null realTargets) should be removed.  See point 1 on #1249 for why.  See also #817 for why the check was there in the first place.

File: src/Stack/Ide.hs (hunk 1 above)
Comment (borsboom) [inline src/Stack/Ide.hs]: @mgsloan: Should it stay with the old behaviour of not starting if the dependencies fail to build, or would the new behaviour of continuing anyway be better?

Comment (mgsloan) [inline src/Stack/Ide.hs]: The new behavior of continuing is better, I think.  At least that way the user ends up with a GHCI session that's somewhat setup.

**Comment by borsboom** [comment]
@mgsloan I've updates based on your comments, please give this a final review and merge if good.

**Comment by mgsloan** [comment]
LGTM!

**Pull request merged**